File: src/app/model/GradeMetric.ts
import { LocationServiceEvent } from './LocationServiceEvent';
import { MetricService } from "../services/metric.service";
import { Util } from "../util/util";
import { Metric, MetricType } from "./Metric";

export class GradeMetric extends Metric {
    private neutralGrade: number = 0;

    private intervalId: any;

    constructor(metricService: MetricService) {
        super(MetricType.GRADE, 'Grade', '%', metricService);
    }

    startLogging(): void {
        if (this.intervalId) return;
        this.intervalId = setInterval(() => {
            const now = Date.now();
            const phoneGrade = this.metricService.getLocationService().gradeForHorizontalPhone;

            this.addValue(phoneGrade - this.neutralGrade, new Date(now));
        }, 500); // Log every half second
    }

    stopLogging(): void {
        if (this.intervalId) {
            clearInterval(this.intervalId);
            this.intervalId = null;
        }
    }

    calibrateGrade(): void {
        const currentGrade = this.metricService.getLocationService().gradeForHorizontalPhone;
        this.neutralGrade = currentGrade;
        console.log(`Calibrated grade to ${this.neutralGrade}%`);
    }

}

File: android/app/src/main/java/com/openbikecomputer/model/GradeMetric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class GradeMetric extends Metric {
    private double neutralGrade = 0;
    private ScheduledExecutorService executorService;
    private boolean started = false;

    public GradeMetric(IMetricService metricService) {
        super(MetricType.GRADE, "Grade", "%", metricService);
    }

    @Override
    public void startLogging() {
        if (started) {
            return;
        }
        started = true;
        executorService = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread thread = new Thread(r, "GradeMetric");
            thread.setDaemon(true);
            return thread;
        });
        executorService.scheduleAtFixedRate(() -> {
            long now = System.currentTimeMillis();
            double phoneGrade = metricService.getLocationService().getGradeForHorizontalPhone();
            addValue(phoneGrade - neutralGrade, now);
        }, 1000, 1000, TimeUnit.MILLISECONDS);
    }

    @Override
    public void stopLogging() {
        if (executorService != null) {
            executorService.shutdownNow();
            executorService = null;
        }
        started = false;
    }

    public void calibrateGrade() {
        neutralGrade = metricService.getLocationService().getGradeForHorizontalPhone();
    }
}
